Show suggestions for mistyped identifiers

Daniel-Cortez opened the pull request:
<!--
Please ensure you have read the CONTRIBUTING.md document before submitting this
pull request. Requests that fail to follow enough of the guidelines will likely
be closed immediately with request to rectify the issues.

Never pull to `master` - always pull to `dev` or a relevant feature branch.
-->

**What this PR does / why we need it**:

Backported this feature from Pawn 3.3; it makes the compiler try to find closest names and suggest them on errors related to undefined (mistyped) identifiers. The current limit for finding close names is 2 mistyped symbols.

Here are some quick examples:
```Pawn
#include <string>
main()
{
    new var = 1;
    printf("%d", val); // error 017: undefined symbol "val"; did you mean "var"?

    printf("%d", celmax); // error 017: undefined symbol "celmax"; did you mean "cellmax"?

    new str[4] = "a";
    strcaf(str, "b"); // error 017: undefined symbol "strcaf"; did you mean "strcat"?

lbl:
    goto lb; // error 019: not a label: "lb"; did you mean "lbl"?
}
```
```Pawn
stock DialogHandler() <dialog:DLG_REGISTER> {}
stock DialogHandler() <dialog:DLG_LOGIN> {}
main()
{
    // error 086: unknown automaton "dialpg"; did you mean "dialog"?
    state dialpg:DLG_REGISTER;

    // error 087: unknown state "DLGLOGIN" for automaton "dialog"; did you mean "DLG_LOGIN"?
    state dialog:DLGLOGIN;
}
```

**Which issue(s) this PR fixes**:

<!--GitHub tip: using "Fixes #<issue number> will automatically close the issue upon being merged-->

**What kind of pull this is**:

<!--Replace [ ] with [x] to mark the checkbox-->

* [ ] A Bug Fix
* [x] A New Feature
* [ ] Some repository meta (documentation, etc)
* [ ] Other

**Additional Documentation**:

<!--
If your PR introduces a change that requires documentation, add it here so it can be added to the wiki.
Feel free to edit the wiki yourself once your PR has been accepted.
-->

Daniel-Cortez commented:
OK, I added tests for this PR. The test code is relatively simple, I just copy-pasted the code from the 1'st comment and added a new case just to make sure the compiler wouldn't output a suggestion if the mistyped name differs from the closest one by more than 2 symbols.

Regarding the other tests, I don't think they need any modifications as this PR already passes all of them just fine.

Southclaws commented:
This one seems quite large, it would be good to get a full code review from an experienced member. Also one of the tests on AppVeyor has apparently got stuck and GitHub provides no link to it for some reason.

Southclaws commented:
Restarted https://ci.appveyor.com/project/Southclaws/compiler/build/172

Y-Less commented on source/compiler/sc5.c:
```diff
@@ -197,6 +198,10 @@ static char *warnmsg[] = {
 };
 
+static char *noticemsg[] = {
+/*001*/  "did you mean \"%s\"?"
```
Would `"; did you mean \"%s\"?\n"` be better?  This would be more in-line with the convention in all the other error/warning messages.

Y-Less commented on source/compiler/sc5.c:
```diff
@@ -244,26 +258,39 @@ static short lastfile;
+    assert(notice>0 && notice<(1+arraysize(noticemsg)));
+    strcpy(string,msg);
+    len=strlen(string);
+    sprintf(&string[len-1],"; %s\n",noticemsg[notice-1]);
```
Were the change to `noticemsg` made, this would become just `strcat`.

Y-Less commented on source/compiler/sc5.c:
```diff
@@ -220,13 +225,22 @@ static int errwarn;
+   * "notice" number
+   */
+  notice=(unsigned long)number >> (sizeof(long)*4);
+  number&=((unsigned long)~0) >> (sizeof(long)*4);
```
Better as `~0UL >> sizeof (long) * 4`?

Daniel-Cortez replied:
I did it this way so there would be more similarity between this line and the above one.

Y-Less replied:
Except casting doesn't sign extend.  So this is:

`(unsigned long)0xFFFFFFFF`

if `sizeof (long) == sizeof (int)` that's fine, if it isn't, you will get `0x00000000FFFFFFFF`.  After shifting that will give 0, not `0xFFFF`.

Daniel-Cortez replied:
Oh, so that's what you meant. Yeah, I already noticed that mistake the first time you left a review comment at that line and fixed it locally by just moving the `~` before `(unsigned long)`. Thanks for pointing it out.

Y-Less commented on source/compiler/sc5.c:
```diff
@@ -423,3 +450,173 @@ int pc_geterrorwarnings()
+  } /* if */
+
+  if (closestname!=NULL) {
+    if (name2!=NULL)
```
Swap the conditions around a `not` condition with an `else` is just confusing.

Y-Less replied:
```c
  if (closestname==NULL) {
    error(number,name,name2);
  } else if (name2==NULL) {
    error(makelong(number,1),name,closestname);
  } else {
    error(makelong(number,1),name,name2,closestname);
  }
```

Y-Less commented on source/compiler/sc5.c:
```diff
@@ -423,3 +450,173 @@ int pc_geterrorwarnings()
+      if (dist<closestdist && dist<=MAX_EDIT_DIST) {
+        closestmatch=ptr;
+        closestdist=dist;
+        if (dist<=1)
```
There's a lot of repetitive code here, which is never great, but I do see why - there are very subte differences in things like the loop condition, for example `find_closeststate` checking the automata is correct so as not to suggest matching states from other ones.

One difference I'm not clear on is this `break`.  It looks like a nice optimisation - if there is a very similar symbol, suggest that and don't bother to continue searching.  However, it is only in this one function, not all of them (which is, of course, the danger with repeated code).

Daniel-Cortez replied:
Yeah, looks like I forgot to add that check in the other functions. Thanks for noting this.

Y-Less commented on source/compiler/sc5.c:
```diff
@@ -423,3 +450,173 @@ int pc_geterrorwarnings()
+  assert(name!=NULL);
+  assert(closestmatch!=NULL);
+  while (ptr!=NULL) {
+    if (ptr->index==fsa && strlen(ptr->name)>0) {
```
It would be nice if `find_closestautomaton` and this could be merged - is there a constant value of `->index` in `sc_automaton_tab`?  If so, `fsa` could be set to that, and an extra function deleted.

Y-Less commented on source/compiler/tests/gh_353.meta:
```diff
@@ -0,0 +1,13 @@
+gh_353.pwn(20) : error 036: empty statement
+gh_353.pwn(23) : error 087: unknown state "DLGLOGIN" for automaton "dialog"; did you mean "DLG_LOGIN"?
+gh_353.pwn(23) : error 036: empty statement
+  """
```
No tests for errors 19, 20, 28, or 80.

Y-Less commented on source/compiler/tests/gh_353.pwn:
```diff
@@ -0,0 +1,26 @@
+	// error 087: unknown state "DLGLOGIN" for automaton "dialog"; did you mean "DLG_LOGIN"?
+	state dialog:DLGLOGIN;
+}
+stock DialogHandler() <dialog:DLG_REGISTER> {}
```
Would also be nice to have a test case something like:

```pawn
Func1() <state_1:STATE_1>{}
Func1() <state_1:STATE_2>{}

Func2() <state_2:BEING_1>{}
Func2() <state_2:BEING_2>{}

state state_2:STATE_2;
```

Which should NOT suggest an alternative, because `STATE_2` is in a different automata.  The code looks like it does this, but confirmation would be nice.

Or even better, in that case suggest that maybe you meant `state_1:STATE_2`, which the code currently does not do.

Daniel-Cortez commented:
> it would be good to get a full code review from an experienced member

By the way, I believe it's worth to note that the code in this PR isn't exactly backported - in fact the only things I copied from Pawn 3.3 were the `noticemsg` array, function `find_closestsymbol_table()` (with a few minor modifications to remove false positives), the Levenshtein distance implementation and maybe a few other simple things (like the `makelong` macro in sc.h and the idea of changing `error(int number,...)` to `error(long number,...)`). All of the other code was either based on the already existing code (f.e. `find_closestautomaton()` and `find_closeststate()` are derived from the code from `find_automaton()` and `find_state()` respectively) or made from scratch.

The original implementation from 3.3 tries to find the closest match even on the first pass which IMO is not the brightest idea because
  1. `error()` doesn't show non-fatal error messages on the first pass anyway (so the same message wouldn't be reported twice);
  2. usually there's a lot of undefined identifiers on the first pass that get resolved by the second pass, so trying to find the closest match for every one of them would be just a waste of time.

And for states and automatons the original code doesn't seem to be well thought out either:
  1. the Levenshtein distance is calculated directly in `find_automaton()` and `find_state()` on every compared name which again slows down compilation for nothing;
  2. instead of having a separate function for suggesting the closest state/automaton name and keeping everything in one place (in sc5.c) the code that tries to find the closest automaton/state names is located in `sc_getstateid()` (sc3.c).

So that's why I partially reimplemented suggestions and kept only the most basic parts of the original code from Pawn 3.3 I mentioned earlier.

Daniel-Cortez commented:
Ok, so I made the following changes to the code:

1. The messages for error 020 were altered for the cases when the user types a keyword instead of an identifier. For example, now when the user mistypes "asset" as "assert", the compiler prints `invalid symbol name "assert"` (before this PR it printed `invalid symbol name ""`).

2. If there is an undefined state name with no close names, the compiler attempts to suggest an automaton with a close name.
  Example:
```Pawn
Func1()<automaton_1:STATE_1>{}
Func2()<automaton_2:BEING_1>{}

state automaton_2:STATE_1; // unknown state "STATE_1" for automaton "automaton_2"; did you mean "automaton_1:STATE_1"?
```
3. Rewrote the tests, added test cases for errors 019, 020 and 080 (there's still no test for error 028; see below).

4. Other minor adjustments.

Things I'm still not sure about:

1. I tried to add a test case for error 028
```Pawn
forward test_e028();
public test_e028()
{
	// error 028: invalid subscript (not an array or too many subscripts): "value"; did you mean "values"?
	const value = 0;
	static const values[2] = { 1, 2 };
	printf("%d\n", value[0]);
	#pragma unused values
}
```
but it made the whole test fail, even though the error output was exactly the same as expected, so I removed that case for now. I suspect the test was failing because of a fatal error (`too many error messages on one line`) triggered by that test case. 
```
gh_353.pwn(81) : error 028: invalid subscript (not an array or too many subscripts): "value"; did you mean "values"?
gh_353.pwn(81) : warning 215: expression has no effect
gh_353.pwn(81) : error 001: expected token: ";", but found "]"
gh_353.pwn(81) : error 029: invalid expression, assumed zero
gh_353.pwn(81) : fatal error 107: too many error messages on one line
```
Is there any way to not trigger a fatal error on 080?

2. Previously if the mistyped symbol (variable/constant/function/label name) was too short (3 or less characters), only a difference of 1 character instead of 2 was allowed, so f.e. the compiler wouldn't suggest "value" for "val". For now I also applied the same rule for automatons and states, but maybe it would be better to remove that rule completely (or rather limit it to the names with length of 2 or less characters instead of 3, so f.e. the compiler would suggest "value" for "val" but wouldn't suggest "var0" for "va")?

@Y-Less

Y-Less commented:
Aren't they all just symbols to the compiler? Though that would be a good thing.

Daniel-Cortez commented:
> Aren't they all just symbols to the compiler?

If there are parentheses after the symbol, the compiler recognizes a function call and thinks the symbol is an unimplemented function.

> Though that would be a good thing.

OK, done. BTW, GitHub says you're still requesting changes - is there anything else that should be done for this PR? (If it's because there's no test case for error 028, I already explained why I couldn't add it.)

Zeex commented:
Merged (https://github.com/pawn-lang/compiler/commit/bc04b2ef048a02a7f4030d807d97a7fc10cb760a).
